launcher/qml/org/prismlauncher/pqcstyle/PQuickStyleItem.cpp

```diff
@@ -0,0 +1,1042 @@
+    , m_sunken(false)
+    , m_raised(false)
+    , m_flat(false)
+    , m_active(true)
+    , m_selected(false)
+    , m_focus(false)
+    , m_hover(false)
+    , m_on(false)
+    , m_horizontal(true)
+    , m_transient(false)
+    , m_sharedWidget(false)
+    , m_minimum(0)
+    , m_maximum(100)
+    , m_value(0)
+    , m_step(0)
+    , m_paintMargins(0)
+    , m_contentWidth(0)
+    , m_contentHeight(0)
+    , m_textureWidth(0)
+    , m_textureHeight(0)
+    , m_focusReason(Qt::NoFocusReason)
+{
+    // Check that we really are a QApplication before attempting to access the
+    // application style.
+    //
+    // Functions used in this file which are safe to access through qApp are:
+    //
+    //   qApp->font() and qApp->fontChanged() - provided by QGuiApplication
+    //   qApp->font("classname") - provided by QApplication but safe
+    //   qApp->layoutDirection() - provided by QGuiApplication
+    //   qApp->wheelScrollLines() - uses styleHints() provided by QGuiApplication
+    //   qApp->testAttribute() and qApp->setAttribute() - provided by QCoreApplication
+    //   qApp->devicePixelRatio() - provided by QGuiApplication
+    //   qApp->palette("classname") - provided by QApplication but safe
+    //
+    // but any use of qApp->style(), even if only trying to test that it exists,
+    // will assert.
+    //
+    // Any use of any other function provided by QApplication must either be checked
+    // to ensure that it is safe to use if not a QApplication, or guarded.
+    if (qobject_cast<QApplication*>(QCoreApplication::instance())) {
+        // We are a QApplication.  Unfortunately there is no styleChanged signal
+        // available, and it only sends QEvent::StyleChange events to all QWidgets.
+        // So watch for the existing application style being destroyed, which means
+        // that a new one is being applied.  See QApplication::setStyle().
+        QStyle* style = qApp->style();
+        if (style) {
+            connect(style, &QObject::destroyed, this, &PQuickStyleItem::styleChanged);
+        }
+    } else if (!s_qstyle) {
+        // We are not a QApplication.  Create an internal copy of the configured
+        // desktop style, to be used for metrics, options and painting.
+        const QString defaultStyleName = QStringLiteral("Fusion");
+        s_qstyle.reset(QStyleFactory::create(defaultStyleName));
+        // Prevent inevitable crashes on nullptr dereference
+        if (!s_qstyle) {
+            qWarning() << "org.prismlauncher.pqcstyle: Could not find QStyle Fusion";
+            ::exit(EXIT_FAILURE);
+        }
+    }
+
+    doResolvePalette();
+    m_palette = new PStylePalette(this);
+
+    m_font = qApp->font();
+    setFlag(QQuickItem::ItemHasContents, true);
+    setSmooth(false);
+
+    qGuiApp->installEventFilter(this);
+}
+
+PQuickStyleItem::~PQuickStyleItem()
+{
+    if (const QStyleOptionButton* aux = qstyleoption_cast<const QStyleOptionButton*>(m_styleoption)) {

```

Comment:
Why is there a need for casting to delete the object?

Reply:
This is what the original code qqc2-desktop-style code does. presumably because there is type information lost in the original pointer that the delete needs?  I thought that information was usually kept by the allocator but I figured I should keep the code as is.

Reply:
Isn't that the hole point of polimorfism? The ability to just associate a pointer of the base and use it as is? Is the destructor somehow not virtual for the casted pointers?

Reply:
well... looking at https://github.com/qt/qtbase/blob/dev/src/widgets/styles/qstyleoption.h#L68
the destructor isn't virtual and never overloaded. so your probably right and I can avoid the cast

Reply:
Yeah no virtual destructor -_-
Leave the casting: it would not destroy the correct object otherwise.